src/GraphRequestUtil.ts

```diff
@@ -58,3 +58,99 @@ export const serializeContent = (content: any): any => {
 	}
 	return content;
 };
+
+export class PathUtils {
+	private static hostRe = /([a-z]+.microsoft.[a-z]+)/g;
+	private static versionRe = /(beta|v1.0)/g;
+	private static fullPathRe = /((v1.0|beta).*?(?=\?|$))/g;
+	private static shortPathRe = /(.*?)((?=\?|$))/g;
+	private static queryRe = /\?.*/g;
+
+	private static getPathFromFullPath(url: string): string {
+		const version = PathUtils.getVersionFrom(url);
+		const path = url.match(PathUtils.fullPathRe);
+		if (!path) {
+			return;
+		}
+
+		return path.shift().replace(version, "");
+	}
+
+	private static getPathFromShortPath(url: string) {
+		const path = url.match(PathUtils.shortPathRe);
+
+		if (!path) {
+			return;
+		}
+
+		return path.shift();
+	}
+
+	/**
+	 * Matches the host from url. Given then url https://graph.microsoft.com/v1.0/me/events
+	 * it will return graph.microsoft.com.
+	 *
+	 * @param {string} url
+	 * @returns {string} host
+	 */
+	public static getHostFrom(url: string): string {
+		const host = url.match(PathUtils.hostRe);
+
+		if (!host) {
+			return "graph.microsoft.com";
+		}
+
+		return host.shift();
+	}
+
+	/**
+	 * Matches the version from url. Given the url https://graph.microsoft.com/v1.0/me/events,
+	 * it will return the version, v1.0 or beta.
+	 *
+	 * @param url
+	 * @returns {string} version
+	 */
+	public static getVersionFrom(url: string): string {
+		const version = url.match(PathUtils.versionRe);
```

Comment:
If a user enters an invalid url. For example -> A simple spelling mistake such as `client.api("https://graph.microsoft.com/bta/me").get()`. In this case the match fails and returns "v1.0 " as a version. It would not be a clear indicator of the error. 
It would be easier to simply forward the incorrect url and let the user see the error when the response fails.  The other advantage of doing this would be that we wouldn't be blocking a valid url which has a different syntax but is marked as invalid by the code.

Reply:
I agree that an approach that is based on url segments rather than values would work better, given that the OData URLs are also mostly based on path segments.

Reply:
@nikithauc what do you think of throwing an error if there's no matching version before the user makes a request?